ocaml/xapi/xapi_pool.ml

```diff
@@ -2073,8 +2073,45 @@ let disable_ssl_legacy = set_ssl_legacy_on_each_host ~value:false
 let enable_ssl_legacy = set_ssl_legacy_on_each_host ~value:true
 
 let set_igmp_snooping_enabled ~__context ~self ~value = 
-  Db.Pool.set_igmp_snooping_enabled ~__context ~self ~value
-  (* To be filled in later. *)
+  if value then
+    Pool_features.assert_enabled ~__context ~f:Features.IGMP_snooping;
+
+  Helpers.assert_using_vswitch ~__context;
+
+  Db.Pool.set_igmp_snooping_enabled ~__context ~self ~value;
+
+  let hosts = Db.Host.get_all ~__context in
+  let networks = Db.Network.get_all ~__context in
+  let host_networks = ref [] in
```

Comment:
The way that the `ref` type and `List.iter` is used here would work, but looks a bit out of place in an OCaml program. If you need to return something from a "loop", then it is more natural to use a "fold".

Also, to return an COULD_NOT_UPDATE_IGMP_SNOOPING_EVERYWHERE error, we just need to know whether one or more of the attach calls has failed, so the return value of the loop can just be a Boolean.

This one is a little more complicated because it a nested loop, but consider something like this (see how the error value can be threaded through):
```
let failed =
  List.fold_left (fun failed host ->
    List.fold_left (fun failed' network ->
      try
        ...attach...;
        failed'
      with _ ->
        true
    ) failed networks
  ) false hosts
in
if failed then
  raise ...
```